zio/tableio/writer.go

```diff
@@ -23,66 +23,74 @@ type Table struct {
 	format    zng.OutFmt
 }
 
-func NewWriter(w io.Writer, flags zio.WriterFlags) *Table {
-	writer := tabwriter.NewWriter(w, 0, 8, 1, ' ', 0)
+func NewWriter(w io.WriteCloser, flags zio.WriterFlags) *Writer {
+	table := tabwriter.NewWriter(w, 0, 8, 1, ' ', 0)
 	var format zng.OutFmt
 	if flags.UTF8 {
 		format = zng.OutFormatZeek
 	} else {
 		format = zng.OutFormatZeekAscii
 	}
-	return &Table{
-		Writer:    w,
+	return &Writer{
+		writer:    w,
 		flattener: zeekio.NewFlattener(resolver.NewContext()),
-		table:     writer,
+		table:     table,
 		limit:     1000,
 		precision: 6,
 		format:    format,
 	}
 }
 
-func (t *Table) writeHeader(typ *zng.TypeRecord) {
+func (w *Writer) writeHeader(typ *zng.TypeRecord) {
 	// write out descriptor headers
 	columnNames := []string{}
 	for _, col := range typ.Columns {
 		//XXX not sure about ToUpper here...
 		columnNames = append(columnNames, strings.ToUpper(col.Name))
 	}
-	fmt.Fprintln(t.table, strings.Join(columnNames, "\t"))
+	fmt.Fprintln(w.table, strings.Join(columnNames, "\t"))
 }
 
-func (t *Table) Write(r *zng.Record) error {
-	r, err := t.flattener.Flatten(r)
+func (w *Writer) Write(r *zng.Record) error {
+	r, err := w.flattener.Flatten(r)
 	if err != nil {
 		return err
 	}
-	if r.Type != t.typ {
-		if t.typ != nil {
-			t.Flush()
-			t.nline = 0
+	if r.Type != w.typ {
+		if w.typ != nil {
+			w.flush()
+			w.nline = 0
 		}
 		// First time, or new descriptor, print header
-		t.writeHeader(r.Type)
-		t.typ = r.Type
+		w.writeHeader(r.Type)
+		w.typ = r.Type
 	}
-	if t.nline >= t.limit {
-		t.Flush()
-		t.writeHeader(t.typ)
-		t.nline = 0
+	if w.nline >= w.limit {
+		w.flush()
+		w.writeHeader(w.typ)
+		w.nline = 0
 	}
 	//XXX only works for zeek-oriented records right now (won't work for NDJSON nested records)
-	ss, changePrecision, err := zeekio.ZeekStrings(r, t.precision, t.format)
+	ss, changePrecision, err := zeekio.ZeekStrings(r, w.precision, w.format)
 	if err != nil {
 		return err
 	}
 	if changePrecision {
-		t.precision = 9
+		w.precision = 9
 	}
-	t.nline++
-	_, err = fmt.Fprintf(t.table, "%s\n", strings.Join(ss, "\t"))
+	w.nline++
+	_, err = fmt.Fprintf(w.table, "%s\n", strings.Join(ss, "\t"))
 	return err
 }
 
-func (t *Table) Flush() error {
-	return t.table.Flush()
+func (w *Writer) flush() error {
+	return w.table.Flush()
+}
+
+func (w *Writer) Close() error {
+	firstErr := w.flush()
+	if err := w.writer.Close(); err != nil && firstErr == nil {
+		firstErr = err
+	}
+	return firstErr
```

Comment:
```suggestion
	err := w.flush()
	if closeErr := w.writer.Close(); closeErr != nil && err == nil {
		err = closeErr
	}
	return err
```

Reply:
This the same but with different names.  I was using Noah's pattern "firstErr" but I have a slight preference for this since the idiomatic "err" stays in the outer scope.

Reply:
btw, microindex.Writer uses the firstErr pattern so we should probably change it to be consistent.  I'll put up another PR.

Reply:
Not that you asked, but I like
```go
	err := w.flush()
	if err2 := w.writer.Close(); err == nil {
		err = err2
	}
	return err

```
for this situation.

Reply:
That's where we landed but do you want me to change the var name `closeErr` to `err2`?

Reply:
https://github.com/brimsec/zq/pull/1236/commits/b803091c1ae2f6295d24547cec707eb33804e901

Reply:
> That's where we landed but do you want me to change the var name closeErr to err2?

Go with your preference.